src/Carousel.js

```diff
@@ -63,233 +100,278 @@ const propTypes = {
-    ) {
-      this.setState({
-        activeIndex: 0,
-        previousActiveIndex: null,
-        direction: null,
-      });
+      return {
+        direction,
+        previousActiveIndex,
+        activeIndex: nextIndex,
+      };
     }
+    return null;
   }
 
-  componentWillUnmount() {
-    clearTimeout(this.timeout);
-    this.isUnmounted = true;
-  }
+  componentDidUpdate(_, prevState) {
+    const { bsPrefix, slide } = this.props;
+    if (
+      !slide ||
+      this.state.activeIndex === prevState.activeIndex ||
+      this._isSliding
+    )
+      return;
 
-  getActiveIndex() {
-    const activeIndexProp = this.props.activeIndex;
-    return activeIndexProp != null ? activeIndexProp : this.state.activeIndex;
-  }
+    const { activeIndex, direction } = this.state;
+    let orderClassName, directionalClassName;
 
-  getDirection(prevIndex, index) {
-    if (prevIndex === index) {
-      return null;
+    if (direction === 'next') {
+      orderClassName = `${bsPrefix}-item-next`;
+      directionalClassName = `${bsPrefix}-item-left`;
+    } else if (direction === 'prev') {
+      orderClassName = `${bsPrefix}-item-prev`;
+      directionalClassName = `${bsPrefix}-item-right`;
     }
 
-    return prevIndex > index ? 'prev' : 'next';
-  }
+    this._isSliding = true;
 
-  handleItemAnimateOutEnd() {
-    this.setState(
-      {
-        previousActiveIndex: null,
-        direction: null,
-      },
-      () => {
-        this.waitForNext();
+    this.pause();
 
-        if (this.props.onSlideEnd) {
-          this.props.onSlideEnd();
-        }
+    // eslint-disable-next-line react/no-did-update-set-state
+    this.safeSetState(
+      { prevClasses: 'active', currentClasses: orderClassName },
+      () => {
+        const items = this.carousel.current.children;
+        const nextElement = items[activeIndex];
+        triggerBrowserReflow(nextElement);
+
+        this.safeSetState(
+          {
+            prevClasses: classNames('active', directionalClassName),
+            currentClasses: classNames(orderClassName, directionalClassName),
+          },
+          () =>
+            transition.end(nextElement, () =>
+              this.safeSetState(
+                { prevClasses: '', currentClasses: 'active' },
+                this.handleSlideEnd,
+              ),
+            ),
+        );
       },
     );
   }
 
-  handleMouseOut() {
-    if (this.isPaused) {
-      this.play();
-    }
+  componentWillUnmount() {
+    clearTimeout(this.timeout);
+    this.isUnmounted = true;
   }
+  safeSetState(state, cb) {
+    if (this.isUnmounted) return;
+    this.setState(state, () => !this.isUnmounted && cb());
+  }
+  handleSlideEnd = () => {
+    const pendingIndex = this._pendingIndex;
+    this._isSliding = false;
+    this._pendingIndex = null;
+
+    if (pendingIndex != null) this.to(pendingIndex);
+    else this.cycle();
+  };
+
+  handleMouseOut = () => {
+    this.cycle();
```

Comment:
only if we're doing "pause on hover"? otherwise why should mousing out reset the timer?